In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9800 non-null   int64  
 1   Order ID       9800 non-null   object 
 2   Order Date     9800 non-null   object 
 3   Ship Date      9800 non-null   object 
 4   Ship Mode      9800 non-null   object 
 5   Customer ID    9800 non-null   object 
 6   Customer Name  9800 non-null   object 
 7   Segment        9800 non-null   object 
 8   Country        9800 non-null   object 
 9   City           9800 non-null   object 
 10  State          9800 non-null   object 
 11  Postal Code    9789 non-null   float64
 12  Region         9800 non-null   object 
 13  Product ID     9800 non-null   object 
 14  Category       9800 non-null   object 
 15  Sub-Category   9800 non-null   object 
 16  Product Name   9800 non-null   object 
 17  Sales          9800 non-null   float64
dtypes: float

## 1. Convert date columns to datetime

In [3]:
df['Order Date'] = pd.to_datetime(df['Order Date'], errors='coerce')
df['Ship Date'] = pd.to_datetime(df['Ship Date'], errors='coerce')

## 2. Drop the 11 NaN

In [4]:
df.dropna(subset=['Postal Code'], inplace=True)

## 3. Feature Engineering: Calculate shipping duration in days

In [5]:
df['Shipping Days'] = (df['Ship Date'] - df['Order Date']).dt.days

## Verify the changes

In [6]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 9789 entries, 0 to 9799
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9789 non-null   int64         
 1   Order ID       9789 non-null   object        
 2   Order Date     3955 non-null   datetime64[ns]
 3   Ship Date      3811 non-null   datetime64[ns]
 4   Ship Mode      9789 non-null   object        
 5   Customer ID    9789 non-null   object        
 6   Customer Name  9789 non-null   object        
 7   Segment        9789 non-null   object        
 8   Country        9789 non-null   object        
 9   City           9789 non-null   object        
 10  State          9789 non-null   object        
 11  Postal Code    9789 non-null   float64       
 12  Region         9789 non-null   object        
 13  Product ID     9789 non-null   object        
 14  Category       9789 non-null   object        
 15  Sub-Category   9789 non-nu

In [7]:
# import pandas as pd
# # 1. Reload your original data first to recover the lost dates
# df = pd.read_csv('Data.csv') 

# # 2. Convert dates by explicitly letting pandas infer the datetime format
# df['Order Date'] = pd.to_datetime(df['Order Date'], dayfirst=False, errors='raise')
# df['Ship Date'] = pd.to_datetime(df['Ship Date'], dayfirst=False, errors='raise')

# # 3. Drop rows with missing Postal Code (as you chose earlier)
# df.dropna(subset=['Postal Code'], inplace=True)

# # 4. Calculate shipping duration in days
# df['Shipping Days'] = (df['Ship Date'] - df['Order Date']).dt.days

# # Verify that Non-Null Count matches the total row count (9789)
# print(df.info())

In [8]:
import pandas as pd

# 1. Reload your original data first to recover the lost dates
df = pd.read_csv('Data.csv') 

# 2. Convert dates using format='mixed' to handle multiple formats automatically
df['Order Date'] = pd.to_datetime(df['Order Date'], format='mixed')
df['Ship Date'] = pd.to_datetime(df['Ship Date'], format='mixed')

# 3. Drop rows with missing Postal Code
df.dropna(subset=['Postal Code'], inplace=True)

# 4. Calculate shipping duration in days
df['Shipping Days'] = (df['Ship Date'] - df['Order Date']).dt.days

# Verify that Non-Null Count matches the total row count (9789)
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 9789 entries, 0 to 9799
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Row ID         9789 non-null   int64         
 1   Order ID       9789 non-null   object        
 2   Order Date     9789 non-null   datetime64[ns]
 3   Ship Date      9789 non-null   datetime64[ns]
 4   Ship Mode      9789 non-null   object        
 5   Customer ID    9789 non-null   object        
 6   Customer Name  9789 non-null   object        
 7   Segment        9789 non-null   object        
 8   Country        9789 non-null   object        
 9   City           9789 non-null   object        
 10  State          9789 non-null   object        
 11  Postal Code    9789 non-null   float64       
 12  Region         9789 non-null   object        
 13  Product ID     9789 non-null   object        
 14  Category       9789 non-null   object        
 15  Sub-Category   9789 non-nu

## First page - Supply chain Dashboard

In [9]:
# 1. Calculate Key Performance Indicators (KPIs)
average_shipping_days = df['Shipping Days'].mean()
maximum_shipping_days = df['Shipping Days'].max()

print("--- Supply Chain Key Metrics ---")
print(f"Average Shipping Time: {average_shipping_days:.2f} days")
print(f"Maximum Shipping Delay (Worst Case): {maximum_shipping_days} days")
print("\n" + "="*40 + "\n")

# 2. Analyze Shipping Days by Ship Mode
ship_mode_analysis = df.groupby('Ship Mode')['Shipping Days'].mean().sort_values(ascending=False)

print("--- Average Shipping Days by Ship Mode ---")
print(ship_mode_analysis)

--- Supply Chain Key Metrics ---
Average Shipping Time: 9.18 days
Maximum Shipping Delay (Worst Case): 214 days


--- Average Shipping Days by Ship Mode ---
Ship Mode
Standard Class    12.105659
First Class        6.594937
Second Class       4.674382
Same Day           0.565056
Name: Shipping Days, dtype: float64


## 🚚 Supply Chain & Operations Insights

### 1. Key Performance Indicators (KPIs)
* **Average Shipping Time:** **9.18 days** — This is significantly high for modern retail/e-commerce standards, indicating a general bottleneck in order fulfillment.
* **Maximum Shipping Delay:** **214 days** 🚨 — A major operational failure. A delay of over 7 months indicates either severe stockouts, lost shipments, or data entry anomalies that require immediate auditing.

---

### 2. Shipping Performance by Mode (`Ship Mode`)
* **Standard Class Inefficiency:** **Standard Class** averages **12.11 days**, which is exceptionally slow. Since this is usually the highest-volume shipping method, it heavily inflates the overall average shipping time.
* **First Class & Second Class Inversion:** Surprisingly, **Second Class (4.67 days)** is outperforming **First Class (6.59 days)**. Customers are paying premium rates for First Class but experiencing longer wait times, which can lead to high dissatisfaction and refund requests.
* **Same Day Success:** **Same Day** delivery is working efficiently, averaging **0.57 days** (less than 14 hours), showing that the fast-track fulfillment pipeline functions properly when prioritized.

In [10]:
# 1. Analyze Shipping Days by Region
region_analysis = df.groupby('Region')['Shipping Days'].mean().sort_values(ascending=False)

print("--- Average Shipping Days by Region ---")
print(region_analysis)
print("\n" + "="*40 + "\n")

# 2. Analyze Shipping Days by State (Top 10 Worst States for Delays)
state_analysis_worst = df.groupby('State')['Shipping Days'].mean().sort_values(ascending=False).head(10)

print("--- Top 10 Worst States by Shipping Delay ---")
print(state_analysis_worst)

--- Average Shipping Days by Region ---
Region
Central    14.776899
East        8.529921
West        6.894268
South       6.844806
Name: Shipping Days, dtype: float64


--- Top 10 Worst States by Shipping Delay ---
State
Wyoming          153.000000
Maine             49.250000
Iowa              40.576923
Idaho             34.666667
Minnesota         32.966292
Utah              32.150943
New Hampshire     26.629630
Wisconsin         26.523810
Nebraska          17.394737
Texas             16.834532
Name: Shipping Days, dtype: float64


### 🌍 Geographic Shipping Performance & Bottlenecks

#### 1. Regional Performance
* **Central Region Crisis:** The **Central** region is the primary operational bottleneck, averaging an alarming **14.78 days** per shipment. This is more than double the shipping times of the **West (6.89 days)** and **South (6.84 days)** regions. Any regional fulfillment centers or carriers operating in the Central zone need an immediate process audit.
* **Efficient Zones:** The **West** and **South** regions are the most efficient, maintaining a relatively healthier average of under 7 days.

#### 2. State-Level Anomalies (Top 10 Bottlenecks)
* **Wyoming Outlier:** **Wyoming** exhibits a catastrophic average delay of **153 days** 🚨. While it might have a low volume of orders, an average this high strongly points to lost transit paths, unfulfilled bulk orders, or severe logistical isolation.
* **The Upper Midwest & Northeast Delays:** **Maine (49.25 days)**, **Iowa (40.58 days)**, **Idaho (34.67 days)**, and **Minnesota (32.97 days)** all show massive delays exceeding a full month. 
* **High-Volume Risk (Texas):** While **Texas** is 10th on the worst-performing list, its average of **16.83 days** is highly critical because Texas is traditionally a high-volume sales state. A delay this high there impacts a large segment of the customer base.

## Save the cleaned dataframe to a new CSV file for Power BI

In [11]:
#df.to_csv('Cleaned_Data.csv', index=False)

## 👥 Marketing & Customer Analytics Insights

### 1. Customer Lifetime Value (CLV) & "Whales" Identification
* **The Ultimate Whale:** **Sean Miller** (`SM-20320`) is the highest-value customer for the company, with a massive lifetime spend of **$25,043.05**. 
* **Top Tier Contributors:** Customers like **Tamara Chand ($19,052.22)** and **Raymond Buch ($15,117.34)** follow closely. These top 5 individuals represent critical high-value accounts that require dedicated relationship management and exclusive VIP offers to ensure retention.

In [12]:
# 1. Calculate Customer Lifetime Value (CLV) as Total Sales per Customer
clv_df = df.groupby(['Customer ID', 'Customer Name'])['Sales'].sum().reset_index()
clv_df.rename(columns={'Sales': 'CLV_Total_Sales'}, inplace=True)

### 2. Customer Segmentation Strategy (`pd.cut` / Logic)
* **The Revenue Backbone (Regulars):** The majority of the customer base falls into the **Regular** segment (**501 customers** spending between $1k and $5k). This is a healthy sign, showing a strong, stable middle-class customer layer that generates consistent revenue.
* **The Core Asset (VIPs):** There are **113 VIP customers** (spending > $5k). Although they represent a smaller portion of the total customer count, their high CLV likely contributes disproportionately to total profitability. 
* **The Growth Opportunity (Occasionals):** There are **179 Occasional customers** (spending < $1k). Marketing should target this specific segment with re-engagement campaigns, bundle discounts, or first-purchase incentives to convert them into "Regular" buyers.

In [13]:
# 2. Customer Segmentation based on CLV thresholds
# VIP: > 5000 | Regular: 1000 - 5000 | Occasional: < 1000
def assign_segment(sales):
    if sales > 5000:
        return 'VIP'
    elif 1000 <= sales <= 5000:
        return 'Regular'
    else:
        return 'Occasional'

clv_df['Customer_Segment'] = clv_df['CLV_Total_Sales'].apply(assign_segment)

# 3. Count how many customers are in each segment
segment_counts = clv_df['Customer_Segment'].value_counts()

print("--- Top 5 Whales (Highest CLV) ---")
print(clv_df.sort_values(by='CLV_Total_Sales', ascending=False).head(5))
print("\n" + "="*50 + "\n")

print("--- Customer Segment Counts ---")
print(segment_counts)

--- Top 5 Whales (Highest CLV) ---
    Customer ID  Customer Name  CLV_Total_Sales Customer_Segment
700    SM-20320    Sean Miller        25043.050              VIP
741    TC-20980   Tamara Chand        19052.218              VIP
621    RB-19360   Raymond Buch        15117.339              VIP
730    TA-21385   Tom Ashbrook        14595.620              VIP
6      AB-10105  Adrian Barton        14473.571              VIP


--- Customer Segment Counts ---
Customer_Segment
Regular       501
Occasional    179
VIP           113
Name: count, dtype: int64


### 🎯 Product Preference Analysis by Segment

#### 1. VIP Segment Preferences (The Tech Buyers)
* **High-Ticket Technology:** The **VIP** segment heavily drives **Technology** sales ($394,226.71). 
* **Top Sub-Categories:** Their absolute favorites are **Machines ($138,012.75)** and **Phones ($114,545.55)**, followed by **Chairs ($111,135.64)**. This proves that VIPs are purchasing high-value infrastructure and heavy enterprise equipment. Marketing should target them with tech upgrades and premium enterprise contracts.

#### 2. Regular Segment Preferences (The Infrastructure Backbone)
* **Balanced Core Buying:** The **Regular** segment spends almost equally across all categories, with **Furniture ($459,023.07)** leading slightly.
* **Top Sub-Categories:** Their top purchases are **Chairs ($199,089.30)** and **Phones ($197,590.26)**. This segment represents steady office setups and standard business operations.

#### 3. Occasional Segment Preferences (The Essentials Buyers)
* **Low-Cost Office Supplies:** **Occasional** buyers focus primarily on **Office Supplies ($42,606.81)**, which are low-ticket, day-to-day items.
* **Top Sub-Categories:** While **Phones** and **Chairs** appear in their top 3, their overall volumes are very low. They mostly come to buy basic essentials. Cross-selling Tech bundles could help upgrade them to the Regular tier.

In [14]:
# 1. Merge customer segments back into the main DataFrame
df_segmented = df.merge(clv_df[['Customer ID', 'Customer_Segment']], on='Customer ID', how='left')

# 2. Analyze Product Categories by Segment
category_pref = df_segmented.groupby(['Customer_Segment', 'Category'])['Sales'].sum().reset_index()
category_pref = category_pref.sort_values(['Customer_Segment', 'Sales'], ascending=[True, False])

print("--- Category Sales by Customer Segment ---")
print(category_pref)
print("\n" + "="*50 + "\n")

# 3. Analyze Top 3 Sub-Categories for each Segment to find deep preferences
print("--- Top 3 Sub-Categories by Segment ---")
subcat_pref = df_segmented.groupby(['Customer_Segment', 'Sub-Category'])['Sales'].sum().reset_index()

for segment in ['VIP', 'Regular', 'Occasional']:
    print(f"\n[ {segment} Customer Preferences ]")
    top_3_sub = subcat_pref[subcat_pref['Customer_Segment'] == segment].sort_values(by='Sales', ascending=False).head(3)
    print(top_3_sub[['Sub-Category', 'Sales']])

--- Category Sales by Customer Segment ---
  Customer_Segment         Category        Sales
1       Occasional  Office Supplies   42606.8130
0       Occasional        Furniture   31533.4153
2       Occasional       Technology   26317.7000
3          Regular        Furniture  459023.0678
5          Regular       Technology  405311.7070
4          Regular  Office Supplies  404397.1900
8              VIP       Technology  394226.7060
7              VIP  Office Supplies  256208.8210
6              VIP        Furniture  232981.9926


--- Top 3 Sub-Categories by Segment ---

[ VIP Customer Preferences ]
   Sub-Category       Sales
45     Machines  138012.750
47       Phones  114545.554
39       Chairs  111135.638

[ Regular Customer Preferences ]
   Sub-Category       Sales
22       Chairs  199089.296
30       Phones  197590.258
31      Storage  146934.578

[ Occasional Customer Preferences ]
   Sub-Category      Sales
13       Phones  14351.886
5        Chairs  11882.597
14      Storage  10

In [15]:
# Save the updated dataset with customer segments included
#df_segmented.to_csv('Cleaned_Data.csv', index=False)

## 👔 Executive Insights (High-Level Report)

### 1. High-Level Executive Metrics
* **Total Sales Revenue:** The company has generated a total revenue of **$2,252,607.41** across all historical data.
* **Operational Volume:** This revenue was driven by **4,916 unique orders** placed by a loyal base of **793 unique customers**.

---

### 2. Strategic Growth Trend (Yearly Performance)
* **The 2016 Dip:** Sales experienced a slight contraction in 2016, dropping from **$479,856.21** (2015) to **$454,315.91** (-5.3%).
* **The Post-2017 Explosion:** The business recovered aggressively in 2017 with a **31.4% growth spike** ($597,225.49) and kept the momentum surging into 2018, hitting an all-time high of **$721,209.81**. This indicates highly successful scaling, marketing, or expansion strategies implemented in the last two years.

---

### 3. Market Geography & Product Breakdown
* **Regional Powerhouses:** The **West ($710,219.68)** and **East ($660,589.36)** regions are the primary revenue engines, collectively contributing over 60% of total sales. The **South** remains the smallest market at **$389,151.46**.
* **Balanced Product Mix:** **Technology** leads the company's portfolio at **$825,856.11**, but **Furniture ($723,538.48)** and **Office Supplies ($703,212.82)** follow extremely closely. This balanced distribution means the company does not rely on a single category to survive, reducing overall market risk.

In [16]:
# Ensure Order Date is in datetime format
df['Order Date'] = pd.to_datetime(df['Order Date'])

# 1. High-Level Executive Metrics
total_sales = df['Sales'].sum()
total_orders = df['Order ID'].nunique()
total_customers = df['Customer ID'].nunique()

print("--- Executive High-Level Metrics ---")
print(f"Total Sales: ${total_sales:,.2f}")
print(f"Total Unique Orders: {total_orders:,}")
print(f"Total Unique Customers: {total_customers:,}")
print("\n" + "="*50 + "\n")

# 2. Regional Performance
region_sales = df.groupby('Region')['Sales'].sum().sort_values(ascending=False).reset_index()
print("--- Sales by Region ---")
print(region_sales)
print("\n" + "="*50 + "\n")

# 3. Category Performance
category_sales = df.groupby('Category')['Sales'].sum().sort_values(ascending=False).reset_index()
print("--- Sales by Product Category ---")
print(category_sales)
print("\n" + "="*50 + "\n")

# 4. Yearly Sales Trend (To check overall growth)
df['Year'] = df['Order Date'].dt.year
yearly_trend = df.groupby('Year')['Sales'].sum().reset_index()
print("--- Yearly Sales Trend ---")
print(yearly_trend)

--- Executive High-Level Metrics ---
Total Sales: $2,252,607.41
Total Unique Orders: 4,916
Total Unique Customers: 793


--- Sales by Region ---
    Region        Sales
0     West  710219.6845
1     East  660589.3560
2  Central  492646.9132
3    South  389151.4590


--- Sales by Product Category ---
          Category        Sales
0       Technology  825856.1130
1        Furniture  723538.4757
2  Office Supplies  703212.8240


--- Yearly Sales Trend ---
   Year        Sales
0  2015  479856.2081
1  2016  454315.9054
2  2017  597225.4900
3  2018  721209.8092
